sui_types/src/batch.rs

```diff
@@ -49,39 +65,74 @@ impl AuthorityBatch {
     /// The first batch for any authority indexes at zero
     /// and has zero length.
     pub fn initial() -> AuthorityBatch {
-        let to_hash = TransactionBatch(Vec::new());
-        let transactions_digest = sha3_hash(&to_hash);
+        let transaction_batch = TransactionBatch(Vec::new());
+        let transactions_digest = sha3_hash(&transaction_batch);
 
         AuthorityBatch {
             next_sequence_number: 0,
             initial_sequence_number: 0,
             size: 0,
             previous_digest: None,
             transactions_digest,
+            transaction_batch,
         }
     }
 
     /// Make a batch, containing some transactions, and following the previous
     /// batch.
     pub fn make_next(
         previous_batch: &AuthorityBatch,
-        transactions: &[(TxSequenceNumber, TransactionDigest)],
+        transactions: &[(
+            TxSequenceNumber,
+            TransactionDigest,
+            Option<Box<TransactionInfoResponse>>,
+        )],
     ) -> AuthorityBatch {
         let transaction_vec = transactions.to_vec();
         debug_assert!(!transaction_vec.is_empty());
 
         let initial_sequence_number = transaction_vec[0].0 as u64;
         let next_sequence_number = (transaction_vec[transaction_vec.len() - 1].0 + 1) as u64;
 
-        let to_hash = TransactionBatch(transaction_vec);
-        let transactions_digest = sha3_hash(&to_hash);
+        let transaction_batch = TransactionBatch(transaction_vec);
+        let transactions_digest = sha3_hash(&transaction_batch);
 
         AuthorityBatch {
             next_sequence_number,
             initial_sequence_number,
             size: transactions.len() as u64,
             previous_digest: Some(previous_batch.digest()),
             transactions_digest,
+            transaction_batch,
+        }
+    }
+
+    /// Make a batch, containing some transactions, and following the previous
+    /// batch, using the digest of the previous batch.
+    pub fn make_next_with_previous_digest(
+        previous_batch_digest: Option<BatchDigest>,
+        transactions: &[(
+            TxSequenceNumber,
+            TransactionDigest,
+            Option<Box<TransactionInfoResponse>>,
+        )],
+    ) -> AuthorityBatch {
+        let transaction_vec = transactions.to_vec();
+        debug_assert!(!transaction_vec.is_empty());
+
+        let initial_sequence_number = transaction_vec[0].0 as u64;
+        let next_sequence_number = (transaction_vec[transaction_vec.len() - 1].0 + 1) as u64;
+
+        let transaction_batch = TransactionBatch(transaction_vec);
+        let transactions_digest = sha3_hash(&transaction_batch);
```

Comment:
This digest now will be different depending on whether the transactions include the `TransactionInfoResponse` or not. So a signature on the batch may fail to validate for this reason, if it was constructed on a batch without TransactionInfoResponse and then it is checked on a batch with them.